### PR TITLE
enable proxy settings for kubernetes build and update docs

### PR DESCRIPTION
**Pull request by mfanjie**
I am in China and badly gcr.io is blocked by GWF hence I need to setup my own proxy to make Kubernetes build, however when I refer build/README.md and follow the guideline to setup proxy in build/build-image/Dockerfile, I still cannot successfully make my own build.

Checking the build logs and source code, I see requires to connect to gcr.io to pull docker images during build time, hence this requires docker machine kube-dev having the capabilities of bypassing the proxy.

I updated the build/README.md to reflect this requirement, so that a user behind a proxy won't be blocked for k8s build. Please kindly review and suggest if this is fine to merge. Thanks.

------------error logs before setting proxy for kube-dev------------------
+++ [0128 11:21:08] Starting Docker build for image: kube-proxy
unable to ping registry endpoint https://gcr.io/v0/
v2 ping attempt failed with error: Get https://gcr.io/v2/: dial tcp 64.233.189.82:443: i/o timeout
v1 ping attempt failed with error: Get https://gcr.io/v1/_ping: dial tcp 64.233.189.82:443: i/o timeout
!!! Error in build/../build/common.sh:774
'"${DOCKER[@]}" build -q -t "${docker_image_tag}" ${docker_build_path} > /dev/null' exited with status 1
Call stack:
1: build/../build/common.sh:774 kube::release::create_docker_images_for_server(...)
2: build/../build/common.sh:727 kube::release::package_server_tarballs(...)
3: build/../build/common.sh:661 kube::release::package_tarballs(...)
4: build/release.sh:40 main(...)
Exiting with status 1
!!! [0128 11:22:24] previous tarball phase failed
!!! Error in build/release.sh:40
'return 1' exited with status 1
Call stack:
1: build/release.sh:40 main(...)
Exiting with status 1
make: **\* [quick-release] Error 1

### CODE REVIEW

**Comment by ixdy** [comment]
I don't have an OS X installation to test on, but would doing something like editing `/var/lib/boot2docker/profile` as described [here](http://mflo.io/2015/08/13/docker-machine-behind-proxy/) be better, since it would hopefully affect all containers?

If that doesn't work, I'd argue we could probably just update `build/common.sh` to update the referenced lines to something like
```
docker-machine create --driver "${DOCKER_MACHINE_DRIVER}" \
  --engine-env HTTP_PROXY="${http_proxy:-}" \
  --engine-env HTTPS_PROXY="${https_proxy:-}" \
  "${DOCKER_MACHINE_NAME}" > /dev/null || {
```

Then, you just need to set the (fairly standard?) `http_proxy` in your environment, rather than editing the build script.

cc @zmerlynn @ihmccreery @david-mcmahon

**Comment by ixdy** [comment]
If we update the script, we should probably also include `--engine-env NO_PROXY="${no_proxy:-}"`.

**Comment by mfanjie** [comment]
@ixdy There is no global proxy configuration can be applied to all instances for dockertoolbox. /var/lib/boot2docker/profile is on the VM being created and it is different for every individual machine. 
The link you posted can be summarized as
1. docker-machine create --driver virtualbox kube-dev
2. docker-machine ssh kube-dev
3. vi /var/lib/boot2docker/profile
4. docker-machine restart kube-dev
Currently, even we doc this solution. The first build will definite fail as kube-dev is created as the very first step of whole build process.

Command docker-machine create with proxy setting is doing the same thing in single command and after kube-dev is created, and when I check /var/lib/boot2docker/profile on kube-dev, the proxy setting is there.

I agree with your suggestion about change the code and use environment variable to define the proxy, but I think it should not use the host proxy settings, for example I do not want to enable global proxy on my Mac, but I need it is configured on kube-dev, agree?

I can make the code change as well the doc change if you feel comfortable on the approach.

**Comment by mfanjie** [comment]
I made changes to support define environment variables for kubernetes build proxy, things will be messed if we use system global proxy setting, for example, if we use http_proxy, we need to define no_proxy to include ip of kube-dev virtual machine, and we need to notify users to check the ip of kube-dev and set it in no_proxy settings when never users rebuild the VM.

My thought is we should provide different proxy settings for kubernetes build rather than OS. e.g. I never set system-wides proxy on my MAC, so I have choice to enable proxy or not for each app.  

This patch is tested on both MAC and Linux, with proxy or without proxy settings.

One thing I want to high light is that to make logic simple it sets proxy entry in /var/lib/boot2docker/profile and set ENV on kube-build Dockerfile even the value is not set.

Like if user did not set proxy, then cat /var/lib/boot2docker/profile
1. docker@kube-dev:~$ cat /var/lib/boot2docker/profile 

EXTRA_ARGS='
--label provider=virtualbox

'
CACERT=/var/lib/boot2docker/ca.pem
DOCKER_HOST='-H tcp://0.0.0.0:2376'
DOCKER_STORAGE=aufs
DOCKER_TLS=auto
SERVERKEY=/var/lib/boot2docker/server-key.pem
SERVERCERT=/var/lib/boot2docker/server.pem

export "HTTP_PROXY="
export "HTTPS_PROXY="
export "NO_PROXY=127.0.0.1"

2. cat _output/images/kube-build\:build-403db1644c/Dockerfile 
ENV http_proxy ""
ENV https_proxy ""
ENV no_proxy 127.0.0.1

File: build/common.sh [hunk 1]
```diff
@@ -203,6 +207,18 @@ function kube::build::is_osx() {
...
 }
 
+function kube::build::update_dockerfile() {
+  if kube::build::is_osx; then
+    sed_arg=\"\"
+  else
+    sed_arg=""
+  fi
+  sed -i $sed_arg "s/KUBE_BUILD_IMAGE_CROSS/${KUBE_BUILD_IMAGE_CROSS}/" ${build_context_dir}/Dockerfile
+  sed -i $sed_arg "s#KUBE_BUILD_HTTP_PROXY#"`echo ${KUBE_BUILD_HTTP_PROXY:-\"\"}`"#" ${build_context_dir}/Dockerfile
```
Comment (ixdy) [inline build/common.sh]: I think you don't need to echo - just `"s#KUBE_BUILD_HTTP_PROXY#${KUBE_BUILD_HTTP_PROXY:-\"\"}#"` should work.

File: build/common.sh [hunk 2]
```diff
@@ -203,6 +207,18 @@ function kube::build::is_osx() {
   [[ "$(uname)" == "Darwin" ]]
 }
 
+function kube::build::update_dockerfile() {
+  if kube::build::is_osx; then
```
Comment (ixdy) [inline build/common.sh]: maybe slightly rework this, just to be consistent with the rest of this file:
```shell
if kube::build::is_osx; then
  local sed_opts=(-i \"\")
else
  local sed_opts=(-i)
fi
sed -i ${sed_opts[@]} "s/...
```

**Comment by ixdy** [comment]
Other than minor comments, this looks fine to me. @zmerlynn @david-mcmahon thoughts?

**Comment by ixdy** [comment]
ok to test

File: build/common.sh [hunk 3]
```diff
@@ -472,11 +488,8 @@ function kube::build::build_image() {
   kube::version::save_version_vars "${build_context_dir}/kube-version-defs"
 
   cp build/build-image/Dockerfile ${build_context_dir}/Dockerfile
-  if kube::build::is_osx; then
-    sed -i "" "s/KUBE_BUILD_IMAGE_CROSS/${KUBE_BUILD_IMAGE_CROSS}/" ${build_context_dir}/Dockerfile
```
Comment (david-mcmahon) [inline build/common.sh]: I'm not osx-savvy.  WHat's up with the 'sed ""'?

Comment (david-mcmahon) [inline build/common.sh]: Oh I see.  That's cute, OSX.   Better to not dup the regex.  Do something like ixdy mentions above.

Comment (ixdy) [inline build/common.sh]: yeah, the original `sed` fix was in https://github.com/kubernetes/kubernetes/pull/20147. I don't know whether to blame OSX/BSD or GNU for being different.

In any case, this PR cleans it up a bit.

File: build/common.sh (hunk 2 above)
Comment (zmerlynn) [inline build/common.sh]: Another option is just to remove `sed -i` and find a place for an outfile, then do a mv. That's the old POSIX way, before `-i` existed.

File: build/common.sh (hunk 3 above)
Comment (david-mcmahon) [inline build/common.sh]: I don't see an easy way out here to coexist with OSX.

This claims to work fine.  I can't test:
sed -i'' 's/old/new/g' *

Using a global SED_ARGS="-i''" could also work or a kube::sed.

Sadly, none of these options allows you to work seamlessly without having to remember to do *something* different.

Comment (ixdy) [inline build/common.sh]: see https://github.com/kubernetes/kubernetes/pull/20147#discussion_r50938658 from the previous PR. it's known working as broken.

File: build/common.sh (hunk 2 above)
Comment (mfanjie) [inline build/common.sh]: @ixdy I saw this suggestion from you before I made the fix.
On GNU linux it works, if we echo the command we can see the command is changed to sed -i ""; however on BSD ${sed_opts[@]} will drop the "", if we echo the command, it prints sed -i xxxx, I do not get a way to print the "" as array element, that's why I changed it to current solution.

Comment (mfanjie) [inline build/common.sh]: LM-SHC-16501199:work fmeng$ uname -a
Darwin LM-SHC-16501199 15.3.0 Darwin Kernel Version 15.3.0: Thu Dec 10 18:40:58 PST 2015; root:xnu-3248.30.4~1/RELEASE_X86_64 x86_64
LM-SHC-16501199:work fmeng$ sed_opts=(-i \"\")
LM-SHC-16501199:work fmeng$ echo $sed_opts
-i
LM-SHC-16501199:work fmeng$

Comment (david-mcmahon) [inline build/common.sh]: You're only getting the first element by referencing $sed_opts.
Try ${sed_opts[@]}

Also try single quotes in the array element instead:
sed_opts=("-i ''")

Comment (mfanjie) [inline build/common.sh]: ah, thanks for pointing this out, I omitted the [@], will make the change.

File: build/common.sh (hunk 3 above)
Comment (mfanjie) [inline build/common.sh]: all comments addressed

File: build/common.sh (hunk 2 above)
Comment (ixdy) [inline build/common.sh]: hm, I'm not sure that `("-i ''")` is correct - that's just treating it as a single array element.

Comment (mfanjie) [inline build/common.sh]: @ixdy It works, I tested on both OSes. 
I wrote simple scripts on my local laptop and verify it before I made the change, and the build succeeded on both OSes. 
Checking my local build output, the  
/Users/fmeng/code/kubernetes/src/k8s.io/validate/kubernetes/_output/images/kube-build:build-b4a36a2a04/Dockerfile
I can see the proxy settings are there.
LM-SHC-16501199:kubernetes fmeng$ more _output/images/kube-build\:build-b4a36a2a04/Dockerfile
Here is the sample without environment variable set.
FROM kube-build:cross-1.5.3-1

MAINTAINER  Joe Beda <jbeda@google.com>

# (set an explicit GOARM of 5 for maximum compatibility)
ENV GOARM 5
ENV GOOS    linux
ENV GOARCH  amd64
ENV http_proxy ""
ENV https_proxy ""
ENV no_proxy 127.0.0.1

**Comment by ixdy** [comment]
LGTM

**Comment by Aiolia1992** [comment]
你好！我在运行docker app时也遇到这个问题了，你是怎么解决的呢

**Pull request merged**